net/raft/raft_test.go

```diff
@@ -116,3 +120,119 @@ func TestStartUninitialized(t *testing.T) {
+	defer nodeA.cleanup()
+	nodeB := newTestNode(t)
+	defer nodeB.cleanup()
+	nodeC := newTestNode(t)
+	defer nodeC.cleanup()
+
+	// Initialize A, creating a fresh cluster.
+	must(t, nodeA.service.Init())
+
+	// Update the cluster to allow A, B and C's addresses.
+	var err error
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeA.addr, "yes"))
+	must(t, err)
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeB.addr, "yes"))
+	must(t, err)
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeC.addr, "yes"))
+	must(t, err)
+
+	// Add B and C to the cluster.
+	must(t, nodeB.service.Join("https://"+nodeA.addr))
+	must(t, nodeC.service.Join("https://"+nodeA.addr))
+
+	// Try setting a value on nodeB.
+	_, err = nodeB.service.Exec(ctx, set("/foo", "bar"))
+	must(t, err)
+
+	// Try reading the value on nodeC's state.
+	must(t, nodeC.service.WaitRead(ctx))
+	got := nodeC.state.Data["/foo"]
+	if got != "bar" {
+		t.Errorf("reading /foo, nodeC got %q want %q", got, "bar")
+	}
+}
+
+func must(t *testing.T, err error) {
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+type testNode struct {
+	dir    string
+	addr   string
+	server *httptest.Server
+	state  *state
+
+	wg      sync.WaitGroup
+	service *Service
+}
+
+func (n *testNode) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	n.wg.Wait()
+	n.service.ServeHTTP(rw, req)
+}
+
+func (n *testNode) cleanup() {
+	os.RemoveAll(n.dir)
+	n.server.Close()
+	// TODO(jackson): stop the Service too
+}
+
+// newTestNode creates a new local raft Service listening on a random
+// port on localhost. It uses the stdlib's httptest package's localhost
+// TLS certificates in both its server and client tls configs. It uses
+// a simple test kv store implementation for the raft Service's state.
+//
+// When finished with a node, call its cleanup method to stop the server
+// and remove its data directory.
+func newTestNode(t *testing.T) *testNode {
+	node := new(testNode)
+	node.wg.Add(1)
+
+	var err error
+	node.dir, err = ioutil.TempDir("", "raft_test.go")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a tls server first so that we can retrieve the address
+	// and tls certificates to pass in to raft.Start.
+	node.server = httptest.NewTLSServer(node)
+	node.addr = node.server.Listener.Addr().String()
+	node.state = newTestState()
+
+	// TODO(jackson): In Go 1.9+ use ts.Client()?:
+	cert, err := x509.ParseCertificate(node.server.TLS.Certificates[0].Certificate[0])
+	if err != nil {
+		node.cleanup()
+		t.Fatal(err)
+	}
+	tlsConfig := net.DefaultTLSConfig()
+	tlsConfig.RootCAs = x509.NewCertPool()
+	tlsConfig.RootCAs.AddCert(cert)
+	tlsConfig.Certificates = node.server.TLS.Certificates
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	// Create the raft service, passing in the server's address
+	node.service, err = Start(node.addr, node.dir, httpClient, node.state)
+	if err != nil {
+		node.cleanup()
+		t.Fatal(err)
+	}
+
+	node.wg.Done()
```

Comment:
Yeah. It's definitely possible to use a channel here. I'm not sure what the correct idiom is here, but I usually think of channels as the right tool when there's data that you want to share between two goroutines or you need to be able to `select { ... }` over multiple asynchronous events, neither is the case here.

I actually copied this use of wait group after the [same thing we do in cored](https://github.com/chain/chain/blob/main/cmd/cored/main.go#L506).

Honestly, maybe I should just drop the synchronization because the node's address hasn't escaped yet, so there shouldn't be any incoming requests. That would change if we support restarting an existing node here, but we can handle it then.